# Connection to database

In [1]:
%load_ext sql

In [2]:
%sql postgresql://appdev@data:5432/appdev

'Connected: appdev@appdev'

In [10]:
%sql SELECT * FROM circuits LIMIT 10;

10 rows affected.


[(1, 'albert_park', 'Albert Park Grand Prix Circuit', 'Melbourne', 'Australia', -37.8497, 144.968, 10, 'http://en.wikipedia.org/wiki/Melbourne_Grand_Prix_Circuit', '(144.968,-37.8497)'),
 (2, 'sepang', 'Sepang International Circuit', 'Kuala Lumpur', 'Malaysia', 2.76083, 101.738, None, 'http://en.wikipedia.org/wiki/Sepang_International_Circuit', '(101.738,2.76083)'),
 (3, 'bahrain', 'Bahrain International Circuit', 'Sakhir', 'Bahrain', 26.0325, 50.5106, None, 'http://en.wikipedia.org/wiki/Bahrain_International_Circuit', '(50.5106,26.0325)'),
 (4, 'catalunya', 'Circuit de Barcelona-Catalunya', 'Montmeló', 'Spain', 41.57, 2.26111, None, 'http://en.wikipedia.org/wiki/Circuit_de_Barcelona-Catalunya', '(2.26111,41.57)'),
 (5, 'istanbul', 'Istanbul Park', 'Istanbul', 'Turkey', 40.9517, 29.405, None, 'http://en.wikipedia.org/wiki/Istanbul_Park', '(29.405,40.9517)'),
 (6, 'monaco', 'Circuit de Monaco', 'Monte-Carlo', 'Monaco', 43.7347, 7.42056, None, 'http://en.wikipedia.org/wiki/Circuit_de_Monaco', '(7.42056,43.7347)'),
 (7, 'villeneuve', 'Circuit Gilles Villeneuve', 'Montreal', 'Canada', 45.5, -73.5228, None, 'http://en.wikipedia.org/wiki/Circuit_Gilles_Villeneuve', '(-73.5228,45.5)'),
 (8, 'magny_cours', 'Circuit de Nevers Magny-Cours', 'Magny Cours', 'France', 46.8642, 3.16361, None, 'http://en.wikipedia.org/wiki/Circuit_de_Nevers_Magny-Cours', '(3.16361,46.8642)'),
 (9, 'silverstone', 'Silverstone Circuit', 'Silverstone', 'UK', 52.0786, -1.01694, None, 'http://en.wikipedia.org/wiki/Silverstone_Circuit', '(-1.01694,52.0786)'),
 (10, 'hockenheimring', 'Hockenheimring', 'Hockenheim', 'Germany', 49.3278, 8.56583, None, 'http://en.wikipedia.org/wiki/Hockenheimring', '(8.56583,49.3278)')]

# Materialized joins assignment

## 1) Describe the content and function of an index

## 2) Analyze a join query on some tables in the f1db schema

## 3) Create a materialized view of it. You will have to

### 1. On the table circuits report:

- What type of indices exists for the table and why they are of that type (and not some other type)

In [28]:
%sql SELECT * FROM information_schema.columns WHERE table_name = 'circuits';

10 rows affected.


[('appdev', 'f1db', 'circuits', 'circuitid', 1, "nextval('circuits_circuitid_seq'::regclass)", 'NO', 'bigint', None, None, 64, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'appdev', 'pg_catalog', 'int8', None, None, None, None, '1', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('appdev', 'f1db', 'circuits', 'circuitref', 2, "''::character varying", 'NO', 'character varying', 255, 1020, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'appdev', 'pg_catalog', 'varchar', None, None, None, None, '2', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('appdev', 'f1db', 'circuits', 'name', 3, "''::character varying", 'NO', 'character varying', 255, 1020, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'appdev', 'pg_catalog', 'varchar', None, None, None, None, '3', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('appdev', 'f1db', 'circuits', 'location', 4, 'NULL::character varying', 'YES', 'character varying', 255, 1020, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'appdev', 'pg_catalog', 'varchar', None, None, None, None, '4', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('appdev', 'f1db', 'circuits', 'country', 5, 'NULL::character varying', 'YES', 'character varying', 255, 1020, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'appdev', 'pg_catalog', 'varchar', None, None, None, None, '5', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('appdev', 'f1db', 'circuits', 'lat', 6, None, 'YES', 'double precision', None, None, 53, 2, None, None, None, None, None, None, None, None, None, None, None, None, None, 'appdev', 'pg_catalog', 'float8', None, None, None, None, '6', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('appdev', 'f1db', 'circuits', 'lng', 7, None, 'YES', 'double precision', None, None, 53, 2, None, None, None, None, None, None, None, None, None, None, None, None, None, 'appdev', 'pg_catalog', 'float8', None, None, None, None, '7', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('appdev', 'f1db', 'circuits', 'alt', 8, None, 'YES', 'bigint', None, None, 64, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'appdev', 'pg_catalog', 'int8', None, None, None, None, '8', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('appdev', 'f1db', 'circuits', 'url', 9, "''::character varying", 'NO', 'character varying', 255, 1020, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'appdev', 'pg_catalog', 'varchar', None, None, None, None, '9', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('appdev', 'f1db', 'circuits', 'position', 10, None, 'YES', 'point', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'appdev', 'pg_catalog', 'point', None, None, None, None, '10', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES')]

In [52]:
%sql SELECT * FROM information_schema.tables WHERE table_schema = 'f1db';

13 rows affected.


[('appdev', 'f1db', 'circuits', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'constructorresults', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'constructors', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'constructorstandings', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'drivers', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'driverstandings', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'laptimes', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'pitstops', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'qualifying', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'races', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'results', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'seasons', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None),
 ('appdev', 'f1db', 'status', 'BASE TABLE', None, None, None, None, None, 'YES', 'NO', None)]

In [54]:
%sql SELECT * FROM pg_indexes WHERE tablename = 'circuits';

3 rows affected.


[('f1db', 'circuits', 'idx_17102_url', None, 'CREATE UNIQUE INDEX idx_17102_url ON circuits USING btree (url)'),
 ('f1db', 'circuits', 'circuits_position_idx', None, 'CREATE INDEX circuits_position_idx ON circuits USING gist ("position")'),
 ('f1db', 'circuits', 'idx_17102_primary', None, 'CREATE UNIQUE INDEX idx_17102_primary ON circuits USING btree (circuitid)')]

As seen in the above segment there exists three indices; "idx_17102_url", "circuits_position_idx" and "idx_17102_primary".

The first one is of the type btree, a default index used to look up URL's. Since it's only for URL's, which are simple values(numbers and strings), it's fine to just use a btree index (O(log(n))).

the second one is of the type gist, a generalised search tree used to build whichever type of tree we choose which includes spatial data. This is a good method when a position is to be determined since we want the geospatial data (longitude and latitude) for that to happen.

The third one is also a type of btree which is used for the circuitid. The reason for using btree here is the same as the first one.

- The amount of space each index takes up

Even though it says it takes up no tablespace, the space the two btrees takes up are 16KB and the gist takes up 8192 bytes

### 2. We are talent scouts looking to win over some of the best new drivers there are. But we don't want them too old! 
### Write a query that finds the winner of all the races, but only if they are younger than 38 years. 
### The query should return the date, driver surname, driver age, track time in miliseconds, race name and circuit name for all races.

In [96]:
%%sql
SELECT  races.name as race_name, 
        races.date as race_date, 
        date_part('year', age(drivers.dob)) AS driver_age,
        milliseconds, 
        drivers.surname, 
        circuits.name as circuit_name FROM results
        JOIN drivers ON date_part('year', age(drivers.dob)) < 38
            AND results.position = 1
            AND drivers.driverid = results.driverid
        JOIN races USING (raceid)
        JOIN circuits ON circuits.circuitid = races.circuitid;

184 rows affected.


[('Australian Grand Prix', datetime.date(2008, 3, 16), 33.0, 5690616, 'Hamilton', 'Albert Park Grand Prix Circuit'),
 ('Bahrain Grand Prix', datetime.date(2008, 4, 6), 36.0, 5466970, 'Massa', 'Bahrain International Circuit'),
 ('Turkish Grand Prix', datetime.date(2008, 5, 11), 36.0, 5209451, 'Massa', 'Istanbul Park'),
 ('Monaco Grand Prix', datetime.date(2008, 5, 25), 33.0, 7242742, 'Hamilton', 'Circuit de Monaco'),
 ('Canadian Grand Prix', datetime.date(2008, 6, 8), 33.0, 5784227, 'Kubica', 'Circuit Gilles Villeneuve'),
 ('French Grand Prix', datetime.date(2008, 6, 22), 36.0, 5510245, 'Massa', 'Circuit de Nevers Magny-Cours'),
 ('British Grand Prix', datetime.date(2008, 7, 6), 33.0, 5949440, 'Hamilton', 'Silverstone Circuit'),
 ('German Grand Prix', datetime.date(2008, 7, 20), 33.0, 5480874, 'Hamilton', 'Hockenheimring'),
 ('Hungarian Grand Prix', datetime.date(2008, 8, 3), 36.0, 5847067, 'Kovalainen', 'Hungaroring'),
 ('European Grand Prix', datetime.date(2008, 8, 24), 36.0, 5732339, 'Massa', 'Valencia Street Circuit'),
 ('Belgian Grand Prix', datetime.date(2008, 9, 7), 36.0, 4979394, 'Massa', 'Circuit de Spa-Francorchamps'),
 ('Italian Grand Prix', datetime.date(2008, 9, 14), 30.0, 5207494, 'Vettel', 'Autodromo Nazionale di Monza'),
 ('Singapore Grand Prix', datetime.date(2008, 9, 28), 36.0, 7036304, 'Alonso', 'Marina Bay Street Circuit'),
 ('Japanese Grand Prix', datetime.date(2008, 10, 12), 36.0, 5421892, 'Alonso', 'Fuji Speedway'),
 ('Chinese Grand Prix', datetime.date(2008, 10, 19), 33.0, 5517403, 'Hamilton', 'Shanghai International Circuit'),
 ('Brazilian Grand Prix', datetime.date(2008, 11, 2), 36.0, 5651435, 'Massa', 'Autódromo José Carlos Pace'),
 ('Malaysian Grand Prix', datetime.date(2007, 4, 8), 36.0, 5534930, 'Alonso', 'Sepang International Circuit'),
 ('Bahrain Grand Prix', datetime.date(2007, 4, 15), 36.0, 5607515, 'Massa', 'Bahrain International Circuit'),
 ('Spanish Grand Prix', datetime.date(2007, 5, 13), 36.0, 5496230, 'Massa', 'Circuit de Barcelona-Catalunya'),
 ('Monaco Grand Prix', datetime.date(2007, 5, 27), 36.0, 6029329, 'Alonso', 'Circuit de Monaco'),
 ('Canadian Grand Prix', datetime.date(2007, 6, 10), 33.0, 6251292, 'Hamilton', 'Circuit Gilles Villeneuve'),
 ('United States Grand Prix', datetime.date(2007, 6, 17), 33.0, 5469965, 'Hamilton', 'Indianapolis Motor Speedway'),
 ('European Grand Prix', datetime.date(2007, 7, 22), 36.0, 7586358, 'Alonso', 'Nürburgring'),
 ('Hungarian Grand Prix', datetime.date(2007, 8, 5), 33.0, 5752991, 'Hamilton', 'Hungaroring'),
 ('Turkish Grand Prix', datetime.date(2007, 8, 26), 36.0, 5202161, 'Massa', 'Istanbul Park'),
 ('Italian Grand Prix', datetime.date(2007, 9, 9), 36.0, 4717806, 'Alonso', 'Autodromo Nazionale di Monza'),
 ('Japanese Grand Prix', datetime.date(2007, 9, 30), 33.0, 7234579, 'Hamilton', 'Fuji Speedway'),
 ('Bahrain Grand Prix', datetime.date(2006, 3, 12), 36.0, 5386205, 'Alonso', 'Bahrain International Circuit'),
 ('Australian Grand Prix', datetime.date(2006, 4, 2), 36.0, 5667870, 'Alonso', 'Albert Park Grand Prix Circuit'),
 ('Spanish Grand Prix', datetime.date(2006, 5, 14), 36.0, 5181759, 'Alonso', 'Circuit de Barcelona-Catalunya'),
 ('Monaco Grand Prix', datetime.date(2006, 5, 28), 36.0, 6223116, 'Alonso', 'Circuit de Monaco'),
 ('British Grand Prix', datetime.date(2006, 6, 11), 36.0, 5151927, 'Alonso', 'Silverstone Circuit'),
 ('Canadian Grand Prix', datetime.date(2006, 6, 25), 36.0, 5677308, 'Alonso', 'Circuit Gilles Villeneuve'),
 ('Turkish Grand Prix', datetime.date(2006, 8, 27), 36.0, 5331082, 'Massa', 'Istanbul Park'),
 ('Japanese Grand Prix', datetime.date(2006, 10, 8), 36.0, 5032413, 'Alonso', 'Suzuka Circuit'),
 ('Brazilian Grand Prix', datetime.date(2006, 10, 22), 36.0, 5513751, 'Massa', 'Autódromo José Carlos Pace'),
 ('Malaysian Grand Prix', datetime.date(2005, 3, 20), 36.0, 5493736, 'Alonso', 'Sepang International Circuit'),
 ('Bahrain Grand Prix', datetime.date(2005, 4, 3), 36.0, 5358531, 'Alonso', 'Bahrain International Circuit'),

### 3. Describe the query using EXPLAIN ANALYZE with at least 5 lines of text.

In [97]:
%%sql
EXPLAIN ANALYZE SELECT  races.name as race_name, 
        races.date as race_date, 
        date_part('year', age(drivers.dob)) AS driver_age,
        milliseconds, 
        drivers.surname, 
        circuits.name as circuit_name FROM results
        JOIN drivers ON date_part('year', age(drivers.dob)) < 38
            AND results.position = 1
            AND drivers.driverid = results.driverid
        JOIN races USING (raceid)
        JOIN circuits ON circuits.circuitid = races.circuitid;

22 rows affected.


[('Hash Join  (cost=82.13..815.57 rows=324 width=66) (actual time=1.208..20.468 rows=184 loops=1)',),
 ('  Hash Cond: (races.circuitid = circuits.circuitid)',),
 ('  ->  Hash Join  (cost=78.48..803.83 rows=324 width=50) (actual time=1.145..20.227 rows=184 loops=1)',),
 ('        Hash Cond: (results.raceid = races.raceid)',),
 ('        ->  Hash Join  (cost=38.53..759.78 rows=324 width=27) (actual time=0.621..19.642 rows=184 loops=1)',),
 ('              Hash Cond: (results.driverid = drivers.driverid)',),
 ('              ->  Seq Scan on results  (cost=0.00..708.96 rows=974 width=24) (actual time=0.007..18.883 rows=974 loops=1)',),
 ('                    Filter: ("position" = 1)',),
 ('                    Rows Removed by Filter: 22703',),
 ('              ->  Hash  (cost=35.03..35.03 rows=280 width=19) (actual time=0.600..0.600 rows=60 loops=1)',),
 ('                    Buckets: 1024  Batches: 1  Memory Usage: 12kB',),
 ('                    ->  Seq Scan on drivers  (cost=0.00..35.03 rows=280 width=19) (actual time=0.023..0.581 rows=60 loops=1)',),
 ("                          Filter: (date_part('year'::text, age((CURRENT_DATE)::timestamp with time zone, (dob)::timestamp with time zone)) < '38'::double precision)",),
 ('                          Rows Removed by Filter: 781',),
 ('        ->  Hash  (cost=27.76..27.76 rows=976 width=39) (actual time=0.513..0.513 rows=976 loops=1)',),
 ('              Buckets: 1024  Batches: 1  Memory Usage: 79kB',),
 ('              ->  Seq Scan on races  (cost=0.00..27.76 rows=976 width=39) (actual time=0.006..0.221 rows=976 loops=1)',),
 ('  ->  Hash  (cost=2.73..2.73 rows=73 width=28) (actual time=0.043..0.043 rows=73 loops=1)',),
 ('        Buckets: 1024  Batches: 1  Memory Usage: 13kB',),
 ('        ->  Seq Scan on circuits  (cost=0.00..2.73 rows=73 width=28) (actual time=0.010..0.022 rows=73 loops=1)',),
 ('Planning time: 0.532 ms',),
 ('Execution time: 20.560 ms',)]

### Answer atleast the following: 

- How many calls are you making?

As seen in the above result it's making 10 calls, which you can see by the arrows and the first "hash join".
Though I don't know why but it's analyzing the table from the buttom up, which for me seems a bit weird.

- How long does it take to perform the query?

It's first has a planning time which takes it 0.532 ms to do, then the execution time takes 20.560 ms.
This can also be seen in the above results.

### 4. Create a materialized view of your query. Using EXPLAIN ANALYZE try to query the view.
### Write at least 5 lines of text explaining what's going on and why the query execution time changed.

In [98]:
%%sql
CREATE MATERIALIZED VIEW race_winners_below_38 AS SELECT  races.name as race_name, 
        races.date as race_date, 
        date_part('year', age(drivers.dob)) AS driver_age,
        milliseconds, 
        drivers.surname, 
        circuits.name as circuit_name FROM results
        JOIN drivers ON date_part('year', age(drivers.dob)) < 38
            AND results.position = 1
            AND drivers.driverid = results.driverid
        JOIN races USING (raceid)
        JOIN circuits ON circuits.circuitid = races.circuitid;

184 rows affected.


[]

In [100]:
%sql EXPLAIN ANALYZE SELECT * FROM race_winners_below_38;

3 rows affected.


[('Seq Scan on race_winners_below_38  (cost=0.00..4.84 rows=184 width=69) (actual time=0.009..0.024 rows=184 loops=1)',),
 ('Planning time: 0.100 ms',),
 ('Execution time: 0.074 ms',)]

The query execution time changed because a materialized view creates a static snapshot of the data. this means it only runs one sequential scan through the results, since we already got it.

Planning and execution time in this is then much lower, as seen planning is on 0.100 ms and Exection is on 0.074 ms. If anything changes in the dataset, updated results won't be shown.